{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "BIO_PROJECT.ipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "id": "u4fvXRf_VQq9",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "5623964c-5391-4e78-efad-7d6cc921f761"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['UACCAGAUGUAUCGACUGUUUGUCGUUAGCCAGCUUAGAGUAUACCAGUGUACUAGCCAGCAAGUUUCAAACGCGGAUC', 'UACCAGGACUGUUUGUCGUGCAUCGUUAGCCAGCUUAGAGUAUACCAGUGUACUAGCCAGCAAGUUUCAAACGCGGAUC', 'UACCAGGACUGUUUGUCGUUAGCCAGCUUAGAGCUCUCCGUAUACCAGUGUACUAGCCAGCAAGUUUCAAACGCGGAUC', 'UACCAGGACUGUUUGUCGUUAGCCAGCUUAGAGUAUACCAGUGUACUAGCCAGCUCGCACAAAGUUUCAAACGCGGAUC']\n",
            "The proteins formed from Exon skipping are: \n",
            "['ASP CYS LEU SER LEU ALA SER LEU GLU TYR THR SER VAL LEU ALA SER LYS PHE GIN THR ARG ILE  ', 'TYR GIN SER LEU ALA SER LEU GLU TYR THR SER VAL LEU ALA SER LYS PHE GIN THR ARG ILE  ', 'TYR GIN ASP CYS LEU VAL TYR GIN CYS THR SER GIN GIN VAL SER ASN ALA ASP  ', 'TYR GIN ASP CYS LEU SER LEU ALA SER LEU GLU SER PHE LYS ARG GLY  ', 'TYR GIN ASP CYS LEU SER LEU ALA SER LEU GLU TYR THR SER VAL LEU ALA SER  ']\n",
            "The proteins formed from 5'or 3' splicing are: \n",
            "['TYR GIN SER LEU ALA SER LEU GLU SER PHE LYS ARG GLY  ', 'ASP CYS LEU VAL TYR GIN CYS THR SER GIN  ']\n",
            "The proteins formed from Intron retention are: \n",
            "['TYR GIN MET TYR ARG LEU PHE VAL VAL SER GIN PRO ARG VAL TYR GIN CYS THR SER GIN GIN VAL SER ASN ALA ASP  ', 'TYR GIN ASP CYS LEU SER CYS ILE VAL SER GIN PRO ARG VAL TYR GIN CYS THR SER GIN GIN VAL SER ASN ALA ASP  ', 'TYR GIN ASP CYS LEU SER LEU ALA SER LEU GLU LEU SER VAL TYR GIN CYS THR SER GIN GIN VAL SER ASN ALA ASP  ', 'TYR GIN ASP CYS LEU SER LEU ALA SER LEU GLU TYR THR SER VAL LEU ALA SER SER HIS LYS VAL SER ASN ALA ASP  ']\n"
          ]
        }
      ],
      "source": [
        "# Alternative-Splicing\n",
        "\n",
        "# input is the DNA string and the introns present in it.\n",
        "# STEP-1: Formatting a fasta file\n",
        "f = open(\"dataset.fasta\", \"r\") \n",
        "lines = f.readlines()\n",
        "DNA_string = \"\"\n",
        "introns = []\n",
        "for x in range(0,len(lines)):\n",
        "  if x==1:\n",
        "    DNA_string = lines[x]\n",
        "    continue\n",
        "  if x%2 != 0:\n",
        "    lines[x] = lines[x].split(\"\\n\")[0]\n",
        "    introns.append(lines[x])\n",
        "\n",
        "\n",
        "# STEP2: Splitting of exons from introns:\n",
        "exons = []\n",
        "for x in introns:\n",
        "  temp = DNA_string.split(x)\n",
        "  exons.append(temp[0])\n",
        "  DNA_string = temp[1]\n",
        "exons.append(temp[1])\n",
        "\n",
        "\n",
        "# STEP-3: Converting DNA to pre-RNA:\n",
        "def dnaToRna(exon):\n",
        "  output = \"\"\n",
        "  for x in exon:\n",
        "    if x == \"G\":\n",
        "      output = output + \"C\"\n",
        "    elif x == \"C\":\n",
        "      output = output + \"G\"\n",
        "    elif x == \"T\":\n",
        "      output = output + \"A\"\n",
        "    elif x == \"A\":\n",
        "      output = output + \"U\"\n",
        "  return output\n",
        "\n",
        "for x in range(0,len(exons)):\n",
        "  exons[x] = dnaToRna(exons[x])\n",
        "for x in range(0,len(introns)):\n",
        "  introns[x] = dnaToRna(introns[x])\n",
        "\n",
        "\n",
        "# STEP-4: ALTERNATIVE SPLICING: 3 METHODS\n",
        "\n",
        "# 1st METHOD: Exon skipping\n",
        "#skipping only one exon\n",
        "firstList = []\n",
        "for x in range(0,len(exons)):\n",
        "  combined = \"\"\n",
        "  for y in range(0,len(exons)):\n",
        "    if y ==x :\n",
        "      continue\n",
        "    combined = combined + exons[y]\n",
        "  firstList.append(combined)\n",
        "\n",
        "\n",
        "# 2nd METHOD: 5' or 3' splicing\n",
        "secondList = []\n",
        "combined1 = \"\"\n",
        "combined2 = \"\"\n",
        "for x in range(0,len(exons)):\n",
        "  if x%2==0:\n",
        "    combined1 = combined1 + exons[x]\n",
        "  if x%2!=0:\n",
        "    combined2 = combined2 + exons[x]\n",
        "secondList.append(combined1)\n",
        "secondList.append(combined2)\n",
        "\n",
        "\n",
        "\n",
        "# 3rd METHOD: Intron retention\n",
        "thirdList = []\n",
        "for i in range(0, len(introns)):\n",
        "    mixup = \"\"\n",
        "    for j in range(0, len(exons)):\n",
        "      mixup = mixup + exons[j];\n",
        "      if(i == j):\n",
        "        mixup = mixup + introns[i]\n",
        "    thirdList.append(mixup)\n",
        "print(thirdList)\n",
        "\n",
        "# STEP-5: Converting spliced rna into proteins\n",
        "table = dict()\n",
        "table[\"phe\"] = [\"UUU\",\"UUC\"]\n",
        "table[\"leu\"] = [\"UUA\",\"UUG\",\"CUU\",\"CUC\",\"CUA\",\"CUG\"]\n",
        "table[\"ile\"] = [\"AUU\",\"AUC\",\"AUA\"]\n",
        "table[\"met\"] = [\"AUG\"]\n",
        "table[\"val\"] = [\"GUU\",\"GUC\",\"GUA\",\"GUG\"]\n",
        "table[\"ser\"] = [\"UCU\",\"UCC\",\"UCA\",\"UCG\",\"AGU\",\"AGC\"]\n",
        "table[\"pro\"] = [\"CUU\",\"CCC\",\"CCA\",\"CCG\"]\n",
        "table[\"thr\"] = [\"ACU\",\"ACC\",\"ACA\",\"ACG\"]\n",
        "table[\"ala\"] = [\"GCU\",\"GCC\",\"GCA\",\"GCG\"]\n",
        "table[\"tyr\"] = [\"UAU\",\"UAC\"]\n",
        "table[\"his\"] = [\"CAU\",\"CAC\"]\n",
        "table[\"gin\"] = [\"CAA\",\"CAG\"]\n",
        "table[\"asn\"] = [\"AAU\",\"AAC\"]\n",
        "table[\"lys\"] = [\"AAA\",\"AAG\"]\n",
        "table[\"asp\"] = [\"GAU\",\"GAC\"]\n",
        "table[\"glu\"] = [\"GAA\",\"GAG\"]\n",
        "table[\"cys\"] = [\"UGU\",\"UGC\"]\n",
        "table[\"trp\"] = [\"UGG\"]\n",
        "table[\"arg\"] = [\"AGA\",\"AGG\",\"CGU\",\"CGC\",\"CGA\",\"CGG\"]\n",
        "table[\"gly\"] = [\"GGU\",\"GGC\",\"GGA\",\"GGG\"]\n",
        "\n",
        "def split_three(sequence):\n",
        "  string = \"\"\n",
        "  count = 0\n",
        "  for i in sequence:\n",
        "    string += i\n",
        "    count += 1\n",
        "    if count ==  3:\n",
        "      string += \" \"\n",
        "      count = 0\n",
        "  return string\n",
        "\n",
        "def returnKeyFromDict(splitted):\n",
        "  protein = \"\"\n",
        "  for x in table:\n",
        "    if splitted in table[x]:\n",
        "      protein = x.upper()\n",
        "  if protein == \"\":\n",
        "    return \"\"\n",
        "  return protein\n",
        "\n",
        "proteinForFirstMethod = []\n",
        "for x in firstList:\n",
        "  first = split_three(x).split(\" \")\n",
        "  outputFirst = \"\"\n",
        "  for y in first:\n",
        "    outputFirst = outputFirst + returnKeyFromDict(y) + \" \"\n",
        "  proteinForFirstMethod.append(outputFirst)\n",
        "\n",
        "proteinForSecondMethod = []\n",
        "for x in secondList:\n",
        "  second = split_three(x).split(\" \")\n",
        "  outputSecond = \"\"\n",
        "  for y in second:\n",
        "    outputSecond = outputSecond + returnKeyFromDict(y) + \" \"\n",
        "  proteinForSecondMethod.append(outputSecond)\n",
        "\n",
        "proteinForThirdMethod = []\n",
        "for x in thirdList:\n",
        "  third = split_three(x).split(\" \")\n",
        "  outputThird = \"\"\n",
        "  for y in third:\n",
        "    outputThird = outputThird + returnKeyFromDict(y) + \" \"\n",
        "  proteinForThirdMethod.append(outputThird)\n",
        "\n",
        "# OUTPUT\n",
        "print(\"The proteins formed from Exon skipping are: \")\n",
        "print(proteinForFirstMethod)\n",
        "print(\"The proteins formed from 5'or 3' splicing are: \")\n",
        "print(proteinForSecondMethod)\n",
        "print(\"The proteins formed from Intron retention are: \")\n",
        "print(proteinForThirdMethod)\n"
      ]
    }
  ]
}